frame/pablo/src/lib.rs

```diff
@@ -467,6 +546,104 @@ pub mod pallet {
 		pub(crate) fn account_id(pool_id: &T::PoolId) -> T::AccountId {
 			T::PalletId::get().into_sub_account(pool_id)
 		}
+		pub(crate) fn do_get_exchange_rate(
+			pool_id: T::PoolId,
+			swap: bool,
+		) -> Result<Rate, DispatchError> {
+			let pair = Self::currency_pair(pool_id)?;
+			let pool_account = Self::account_id(&pool_id);
+			let pair = if swap { pair.swap() } else { pair };
+			let pool_base_asset_under_management =
+				T::Convert::convert(T::Assets::balance(pair.base, &pool_account));
+			let pool_quote_asset_under_management =
+				T::Convert::convert(T::Assets::balance(pair.quote, &pool_account));
+			Ok(Rate::checked_from_rational(
+				pool_base_asset_under_management,
+				pool_quote_asset_under_management,
+			)
+			.ok_or(ArithmeticError::Overflow)?)
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+            let current_timestamp = T::Time::now();
+            // let twap_enabled_pools : Vec<T::PoolId> = PriceCumulativesState::<T>::iter_keys().collect(); 
+            let twap_enabled_pools : Vec<T::PoolId> = TWAPState::<T>::iter_keys().collect(); 
+            twap_enabled_pools.iter().fold(Weight::zero(), |weight, pool_id| {
+                let rate_base = Self::do_get_exchange_rate(*pool_id, false);
+                let rate_quote = Self::do_get_exchange_rate(*pool_id, true);
```

Comment:
We could possibly avoid calculating the rate again here for `rate_quote`, by taking `1/rate_base`, saves some storage calls

Reply:
I tried but code gets ungly specially checking case for None, and Conversion from Rate to Balance etc. So I am not including this in current PR. It is an optimization that can be done later if it really impacts performance.